modules/dnn/src/onnx/onnx_graph_simplifier.cpp

```diff
@@ -154,6 +154,32 @@ class SoftMaxSubgraph : public Subgraph
     int axis;
 };
 
+class FuseGatherCastSubgraph : public Subgraph
+{
+public:
+    FuseGatherCastSubgraph()
+    {
+        int input = addNodeToMatch("");
+        int index = addNodeToMatch("Constant");
+        int gather = addNodeToMatch("Gather", input, index);
+        addNodeToMatch("Cast", gather);
```

Comment:
What if we will just add an entry for `Cast` layer at https://github.com/opencv/opencv/blob/master/modules/dnn/src/onnx/onnx_importer.cpp?

Reply:
But here how can we use that, I mean we are trying to fuse upsample/resize subgraph.

Reply:
Do you mean adding cast layer implementation in [elementwise_layer](https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/elementwise_layers.cpp) ?

Reply:
Got it, thanks!

So it still related to https://github.com/opencv/opencv/issues/16323 and https://github.com/opencv/opencv/issues/14713 which were not actually resolved by https://github.com/opencv/opencv/issues/16323.

I tried to open https://github.com/opencv/opencv/pull/16710 but it seems like it's better to have more generic solution for all three combinations, yes.

So if you don't mind, let's move a test from PyTorch 1.2. to your PR (see https://github.com/opencv/opencv_extra/pull/716).

I'll keep only Mul layer related changes.

Reply:
Sure, I will add the test. 
So you want me to remove Gather-Cast related changes and only keep Mul-Cast related changes ?

Reply:
yeah, you are right, [#16323](https://github.com/opencv/opencv/issues/16323) and [#14713](https://github.com/opencv/opencv/issues/14713) were not completely resolved by [#16573](https://github.com/opencv/opencv/pull/16573) Those models would have worked if exported using recent Pytorch version. Actually, i thought that PR was getting complex so didn't add these modifications there.
So I intend to make those changes generic for all pytorch versions with this PR.